### PR TITLE
Strictly type createError

### PR DESCRIPTION
**Pull request by cm-ayf**
This PR provides better (narrower) type for createError and resulting error constructor.
While index.test-d.ts is changed, it is purely because test suite requires to match type exactly, so it ensures that this PR does not include breaking change.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Review by L2jLiga — APPROVED** [review]
lgtm

File: types/index.d.ts [hunk 1]
```diff
@@ -7,15 +10,14 @@ declare namespace createError {
     statusCode?: number;
   }
 
-  export interface FastifyErrorConstructor {
-    new(a?: any, b?: any, c?: any): FastifyError;
-    (a?: any, b?: any, c?: any): FastifyError;
-    readonly prototype: FastifyError;
+  export interface FastifyErrorConstructor<E extends {} = {}> {
```
Comment (Uzlopak) [inline types/index.d.ts:13]: `extends {} = {}` is afaik not typesafe.
Wont this be a potential issue? What about 
```suggestion
  export interface FastifyErrorConstructor<E extends { code: string; statusCode?: number; } = { code: string: statusCode?: number;}> {
```

Comment (Uzlopak) [reply types/index.d.ts:13]: I tested it and the typings test run through. I think it is better to use my suggestion.

Comment (cm-ayf) [reply types/index.d.ts]: Thank you for suggestion! I added that restriction.

**Review by Uzlopak — APPROVED** [review]
Thank you. Nice PR.

**Pull request merged**